8329820: [Linux] Prefer EGL over GLX

tsayao opened the pull request:
Wayland implementation will require EGL. 

EGL works with Xorg as well. The idea is to be EGL first and if it fails, fallback to GLX. A force flag `prism.es2.forceGLX=true` is available.


See:
[Switching the Linux graphics stack from GLX to EGL](https://mozillagfx.wordpress.com/2021/10/30/switching-the-linux-graphics-stack-from-glx-to-egl/)
[Prefer EGL to GLX for the GL support on X11](https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3540)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (3 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 2 [Authors](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8329820](https://bugs.openjdk.org/browse/JDK-8329820): [Linux] Prefer EGL over GLX (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1381/head:pull/1381` \
`$ git checkout pull/1381`

Update a local copy of the PR: \
`$ git checkout pull/1381` \
`$ git pull https://git.openjdk.org/jfx.git pull/1381/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1381`

View PR using the GUI difftool: \
`$ git pr show -t 1381`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1381.diff">https://git.openjdk.org/jfx/pull/1381.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1381#issuecomment-2041132288)
</details>

lukostyra reviewed (commented):
I took a brief look at your changes and I shared some comments.

Overall the changes look good. The EGL/GLX loading in GLFactory feels a bit hacky, but other than making EGL a completely separate Prism backend (which would duplicate/reorganize a lot of common EGL/GLX code - definitely a no-go IMO) I don't have a better idea how to solve this. Maybe with an ES2-specific property like `prism.es2.forceglx=true`?

This brings me to a question - other than some form of debugging/fallback for when EGL is still new to JFX, do we even need GLX implementation when we have a working EGL implementation? I'm fairly sure that current distros (especially ones officially supported by JFX) have EGL support available at this point in time, and since it's an intermediary layer between application and runtime it shouldn't depend on used GPU.

lukostyra commented:
Also a side note - the PR title does not match the JBS issue.

tsayao commented on modules/javafx.graphics/src/main/native-prism-es2/linux/egl/LinuxGLContext.c:
```diff
@@ -0,0 +1,311 @@
+
+    /* allocate the structure */
+    ctxInfo = (ContextInfo *) malloc(sizeof (ContextInfo));
+    if (ctxInfo == NULL) {
```
Added

tsayao commented on modules/javafx.graphics/src/main/native-prism-es2/linux/egl/LinuxGLContext.c:
```diff
@@ -0,0 +1,311 @@
+
+    /* set function pointers */
+    ctxInfo->glActiveTexture = (PFNGLACTIVETEXTUREPROC)
+            dlsym(RTLD_DEFAULT, "glActiveTexture");
```
Yes - changed.

tsayao commented on modules/javafx.graphics/src/main/java/com/sun/prism/es2/LinuxEGLContext.java:
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2012, 2013, Oracle and/or its affiliates. All rights reserved.
```
Fixef

kevinrushforth commented:
This will need a fair bit of testing and review. I would like a chance to review it, but won't be able to get to it for a couple weeks.

Reviewers: @kevinrushforth @lukostyra @johanvos @arapte 
/reviewers 3

tsayao commented:
I have to figure out how to  test  Monocle, since I have changed the logic of defines on `PrismES2Defs.h`.

lukostyra reviewed (commented):
I left a few comments, mostly for adjusting error-checking. Once those are fixed and others have their input I'll test those changes.

tsayao commented on modules/javafx.graphics/src/main/native-prism-es2/linux/egl/LinuxGLContext.c:
```diff
@@ -0,0 +1,311 @@
+    }
+    interval = (vSyncNeeded) ? 1 : 0;
+    ctxInfo->state.vSyncEnabled = vSyncNeeded;
+    eglSwapInterval(ctxInfo->eglDisplay, interval);
```
Fixed. `eglSwapInterval` fails without a surface, so added a check.

tsayao commented on modules/javafx.graphics/src/main/native-prism-es2/linux/egl/LinuxGLContext.c:
```diff
@@ -0,0 +1,311 @@
+    ctxInfo->glBlitFramebuffer = (PFNGLBLITFRAMEBUFFERPROC)
+            dlsym(RTLD_DEFAULT,"glBlitFramebuffer");
+
+    eglSwapInterval(eglDisplay, 0);
```
Removed this call as it needs a surface.

tsayao commented on modules/javafx.graphics/src/main/native-prism-es2/linux/egl/LinuxGLContext.c:
```diff
@@ -0,0 +1,311 @@
+    initState(ctxInfo);
+
+    // Release context once we are all done
+    eglMakeCurrent(eglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
```
Fixed.

tsayao commented on modules/javafx.graphics/src/main/native-prism-es2/linux/egl/LinuxGLDrawable.c:
```diff
@@ -0,0 +1,120 @@
+        return 0;
+    }
+
+    EGLSurface eglSurface = eglCreateWindowSurface(pfInfo->eglDisplay, pfInfo->eglConfig,
```
Added.

tsayao commented on modules/javafx.graphics/src/main/native-prism-es2/linux/egl/LinuxGLContext.c:
```diff
@@ -0,0 +1,311 @@
+    ctxInfo->glGetProgramInfoLog = (PFNGLGETPROGRAMINFOLOGPROC)
+            eglGetProcAddress("glGetProgramInfoLog");
+    ctxInfo->glTexImage2DMultisample = (PFNGLTEXIMAGE2DMULTISAMPLEPROC)
+            dlsym(RTLD_DEFAULT,"glTexImage2DMultisample");
```
You're right. Fixed.

tsayao commented:
It is possible to set the platform with `EGL_PLATFORM=x11` environment variable. 
Using ` EGL_LOG_LEVEL=debug ` currently outputs that the platform was detected in build time - I think it's wrong.

tsayao commented:
I'll have to fix it

I got a reply explaining it (the docs where not very clear)
https://lists.freedesktop.org/archives/mesa-users/2024-April/001727.html

tsayao commented:
I changed it to use platform specific way to get `EGLDisplay` so `eglGetDisplay` does not have to guess the platform (it can guess wrong).

It can be seen by setting `EGL_LOG_LEVEL=debug` (with `eglGetDisplay` it will output the guessing part).

I also reverted the Dummy window code change.

tsayao commented:
I will no longer commit to this change.

tsayao commented:
Reopening so maybe it could be usefult to [8332108](https://bugs.openjdk.org/browse/JDK-8332108)

tsayao commented:
Reopening so maybe it could be usefult to [8332108](https://bugs.openjdk.org/browse/JDK-8332108)

tsayao commented:
Testing [8332108](https://bugs.openjdk.org/browse/JDK-8332108) with this branch seems to have a significance in favor of EGL:

Using this branch,  compare the attached example with and without `-Dprism.es2.forceGLX=true`

tsayao commented:
It seems to make a difference when display refresh rate is lower (60Hz). With higher refresh (160hz), no difference.

**EGL:**
![Screenshot from 2024-05-26 08-22-04](https://github.com/openjdk/jfx/assets/30704286/325c4e49-6d45-4bd4-8ff8-2aa220d226b4)


**GLX:**
![Screenshot from 2024-05-26 08-22-24](https://github.com/openjdk/jfx/assets/30704286/e5a8167a-cca9-48bc-acaa-231e07363b06)

tsayao commented:
I will get back to this next week.

tsayao commented:
Would it be better to still default to GLX and have a flag to use EGL (and switch within a few releases), or to go with EGL and have a flag to fallback to GLX?

Currently it defaults to EGL first, and if it fails, try GLX (I don't know a scenario where EGL won't work, on Xorg o XWayland).

tsayao commented:
`SetSceneScalingTest` currently fails with `EGL_BAD_ALLOC` on `nCreateDrawable`. I still don't know why - by the file name I would suspect something related to scaling - but there's no scaling on the test (not that I can see).

lukostyra commented:
> by the file name I would suspect something related to scaling - but there's no scaling on the test (not that I can see).

It is related to screen's DPI setting. We had a problem with UI elements sometimes being incorrectly rendered/shifted when scaling setting in the system was set to more than 100% - see [JDK-8299968](https://bugs.openjdk.org/browse/JDK-8299968)

tsayao commented:
EGL requires an `EGLSurface` to be created, and the most obvious place is in `createDrawable`. The problem is that `DrawableInfo` is never freed, so the `SetSceneScalingTest` fails when another Scene is created on the same Stage with `EGL_BAD_ALLOC` - That's because `createDrawable` is called again and it already has a `EGLSurface` associated with it. 
So I need to resolve #1586 first.

tsayao commented:
#1586 was integrated and merged here, so tests now pass because `EglSurface` is properly freed.

It's ready for review.

tsayao commented:
I would appreciate it this can be put on the review queue.

It does seems EGL takes less CPU time on animations, but If anyone has suggestions on how I could provide other comparisons, it would be nice.

kevinrushforth commented:
I am still unsure whether we want to proceed with this at this time. Until we get to the point of knowing which direction we will go for an eventual pure Wayland port, this seems premature. If and when this does go forward, I would rather see EGL initially be done as a fallback, along with a switch to enable it, rather than introducing it and making it the default at the same time. That would be the usual way to introduce a new mode of operation like this.

@lukostyra and @arapte will want to comment as well.

tsayao commented:
Makes sense to only use EGL when enabled, at least until it's mature and known to work for most setups - I can do this change.

Wayland would require EGL, but it works fine with Xorg for quite some time. (I would say Ubuntu 16.04, but I must check it to be sure).

Gtk4 does prefer EGL on X11:
https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gdk/x11/gdkdisplay-x11.c

See `gdk_x11_display_init_gl_backend`.

Gtk3 uses GLX for X11.

Twig6943 commented:
Would be cool if this got merged. X is slowly dying and should be avoided if possible.

tsayao commented:
> Would be cool if this got merged. X is slowly dying and should be avoided if possible.

The Wayland backend is not implemented here, as it would be a significantly larger task.

Twig6943 commented:
@tsayao @Glavo linked me to this issue tho
